File: lib/amrc/factoryplus/service_client/service_interface.py
# Factory+ Python client library
# Service interface
# Copyright 2023 AMRC

from email.message  import Message

from .service_error import ServiceError    

# Apparently this is the best way to do this...
def content_type (res):
    m = Message()
    m.set_default_type("application/octet-stream")
    ct = res.headers.get("Content-Type")
    if ct is not None:
        m["content-type"] = ct
    return m.get_content_type()

class ServiceInterface:
    def __init__ (self, fplus, **opts):
        self.fplus = fplus

    def error (self, msg, status=None):
        raise ServiceError(msg, service=self.service, status=status)

    def fetch (self, url, **opts):
        method = opts.pop("method", "GET")

        headers = opts.pop("headers", {})
        headers["Accept"] = "application/json"

        if "json" in opts:
            headers["Content-Type"] = "application/json"
        elif "content_type" in opts:
            headers["Content-Type"] = opts.pop("content_type")
        elif "data" in opts:
            headers["Content-Type"] = "application/octet-stream"

        res = self.fplus.http.fetch(
            service=self.service,
            url=url, method=method, headers=headers,
            **opts)

        json = None
        if content_type(res) == "application/json":
            json = res.json()

        return (res.status_code, json)

    def ping (self):
        st, ping = self.fetch("ping")
        if st != 200:
            return None
        return ping

File: lib/service/service-interface.js
/*
 * Factory+ NodeJS Utilities
 * Service interface base class.
 * Copyright 2022 AMRC.
 */

import content_type from "content-type";
import Optional from "optional-js";

import { Service } from "../uuids.js";

const Names = Object.fromEntries(
    Object.entries(Service)
        .filter(([k, v]) => k != "Registry")
        .map(([n, u]) => [u, n]));

export class ServiceError extends Error {
    constructor (service, message, status) {
        super(message);
        this.service = service;
        this.status = status;
    }

    toString () {
        const srv = Names[this.service] ?? this.service;
        const st = this.status == undefined ? "" : `: ${this.status}`;
        return `${srv}: ${this.message}${st}`;
    }
}

export class ServiceInterface {
    constructor (fplus) {
        this.fplus = fplus;
        this.debug = fplus.debug;
    }

    throw (message, status) {
        throw new ServiceError(this.service, message, status);
    }

    async fetch (opts) {
        if (typeof opts == "string")
            opts = { url: opts };

        const body = "body" in opts ? JSON.stringify(opts.body) : undefined;

        const headers = { ...opts.headers };
        headers["Accept"] = opts.accept ?? "application/json";
        if (body)
            headers["Content-Type"] = opts.content_type ?? "application/json";

        opts = {
            ...opts,
            service:    this.service,
            headers,
            body,
        };
        const res = await this.fplus.fetch(opts);
        /* ?? here because Optional.or doesn't work properly */
        const json = Optional.of(opts.method ?? "GET")
            .filter(m => m != "HEAD")
            .map(_ => res.headers.get("Content-Type"))
            .map(content_type.parse)
            .filter(ct => ct.type == "application/json")
            .map(_ => res.json())
            .orElse(undefined);

        /* We are only interested in strong etags in the form of a UUID */
        const etag = Optional
            .ofNullable(res.headers.get("ETag"))
            .map(et => 
                /^"([0-9a-f]{8}(-[0-9a-f]{4}){3}-[0-9a-f]{12})"$/.exec(et))
            .map(m => m[1])
            .orElse(undefined);

        return [res.status, await json, etag];
    }

    async ping () {
        const [st, ping] = await this.fetch("/ping");
        if (st != 200) return;
        return ping;
    }
}
